controllers/trivy_vuln_crds_controller.go

```diff
@@ -0,0 +1,219 @@
+			}
+
+			var report v1alpha1.VulnerabilityReport
+			if err := json.Unmarshal(bytes, &report); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			labelsWithProjectMeta := r.appendProjectMetaToReportLabels(report)
+
+			// take report and convert to lagoon problems payload object
+			problemsPayload := updateReportPayloadToLagoonProblems(report)
+
+			var sendData = LagoonInsightsMessage{
+				Payload:     []interface{}{problemsPayload},
+				Annotations: report.Annotations,
+				Labels:      labelsWithProjectMeta,
+			}
+
+			// marshall data ready for sending to broker
+			marshalledData, err := json.Marshal(sendData)
+			if err != nil {
+				log.Error(err, "Unanle to marshall Vulnerability report data")
+				return ctrl.Result{}, err
+			}
+
+			//log.Info(fmt.Sprintf("VulnerabilityReport CRD name: %s", marshalledData))
+			//log.Info(fmt.Sprintf("VulnerabilityReport CRD: %s\n", bytes))
+			//log.Info(fmt.Sprintf("VulnerabilityReport CRD: %s\n", report.ObjectMeta.Name))
+
+			if r.WriteToQueue {
+				if err := r.MessageQWriter(marshalledData); err != nil {
+
+					log.Error(err, "Unable to write to message broker")
+
+					return ctrl.Result{}, err
+				}
+			}
+
+			// add an annotation to the report indicating it has been processed
+			report.Annotations["lagoon.sh/insightsProcessed"] = "true"
+			if err := r.Client.Update(ctx, &report); err != nil {
+				return ctrl.Result{}, err
+			}
+
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func updateReportPayloadToLagoonProblems(report v1alpha1.VulnerabilityReport) interface{} {
+	var problems []map[string]interface{}
+	labels := report.GetLabels()
+
+	for _, problem := range report.Report.Vulnerabilities {
+		var score float64
+		if problem.Score != nil {
+			score = *problem.Score
+		}
+		scoreStr := strconv.FormatFloat(score, 'f', 1, 64)
+		s, _ := strconv.ParseFloat(scoreStr, 64)
+		scoreFloat := s * 0.1
+
+		p := map[string]interface{}{
+			"environment":       labels["lagoon.sh/environmentId"],
+			"identifier":        problem.VulnerabilityID,
+			"associatedPackage": problem.Resource,
+			"version":           problem.InstalledVersion,
+			"fixedVersion":      problem.FixedVersion,
+			"source":            report.Report.Scanner.Name,
+			"service":           report.Report.Artifact.Repository,
+			"data":              "{}",
+			"severity":          problem.Severity,
+			"description": func() string {
+				if problem.Description == "" {
+					return problem.Title
+				}
+				return problem.Description
+			}(),
+			"links":         problem.PrimaryLink,
+			"severityScore": fmt.Sprintf("%.2f", scoreFloat),
+		}
+
+		problems = append(problems, p)
+	}
+
+	return map[string]interface{}{"problems": problems}
+}
+
+func (r *TrivyVulnerabilityReportReconciler) appendProjectMetaToReportLabels(report v1alpha1.VulnerabilityReport) map[string]string {
+	l := report.Labels
+	m := report.ObjectMeta
+
+	if labels, ok := r.PredicateData.(map[string]string); ok {
```

Comment:
So this is what I was worried about - I don't know how we guarantee this PredicateData comes from the currently processed report.

Reply:
Now I've swapped this out for a concurrency-safe `PredicateDataMap`- it _should_ have stored the right data from the report when looking at the current event object. Let's discuss this on our call though.